design/baremetal-operator/configmap_for_host_inspection_data.md

```diff
@@ -0,0 +1,186 @@
+and BareMetalHost must be in the same namespace.
+
+- BareMetalHost is created
+- Operator creates a InspectionData with the same name and namespace as
+    BareMetalHost and sets ownerReferences
+- Provisioning is requested, operator fills out the InspectionData with the
+    host inspection data
+- Pivoted to a target cluster, BMO restores host inspection data from InspectionData
+- Deprovisioning is requested, operator cleans up the inspection data from the
+    InspectionData
+- BareMetalHost deletion requested, operator deletes the InspectionData and
+    BareMetalHost
+
+Operator already respects `baremetalhost.metal3.io/paused` annotation if set on
+a BareMetalHost and stops the reconciliation for the BareMetalHost. This was
+added as part of Cluster API (CAPI) workflow prerequisite, so that if a CAPI Cluster
+object is paused, the provider is capable of pausing all the child objects like
+BareMetalHost. Since InspectionData is dependent on BareMetalHost, we need to
+ensure that when Cluster is paused, we pause reconciliation of InspectionData too.
+
+The spec of InspectionData:
+
+```go
+// Adding HardwareDetails struct here for visibility. During implementation,
+// we don't have to duplicate it, but use from baremetalhost_types.go.
+type HardwareDetails struct {
+
+    SystemVendor HardwareSystemVendor `json:"systemVendor,omitempty"`
+    Firmware     Firmware             `json:"firmware,omitempty"`
+    RAMMebibytes int                  `json:"ramMebibytes,omitempty"`
+    NIC          []NIC                `json:"nics,omitempty"`
+    Storage      []Storage            `json:"storage,omitempty"`
+    CPU          CPU                  `json:"cpu,omitempty"`
+    Hostname     string               `json:"hostname,omitempty"`
+
+}
+
+type InspectionDataSpec struct {
+    // The hardware discovered on the host.
+    HardwareDetails *HardwareDetails `json:"hardware,omitempty"`
+}
+```
+
+The status of InspectionData:
+
+```go
+type InspectionDataStatus struct {
+    // Last time the status was updated/observed.
+    // +optional
+    LastUpdated *metav1.Time `json:"lastUpdated,omitempty"`
+}
+```
+
+### Risks and Mitigations
+
+Any suggestions?
+
+### Work Items
+
+- Add HardwareDetails API
+- Modify the operator code
+- Clean up statusAnnotation related code from BMO and CAPM3
+- Add unit tests
+- Update documentation
+
+### Dependencies
+
+None
+
+### Test Plan
+
+Unit tests should be added. Integration test doesn't require any change.
+
+### Upgrade / Downgrade Strategy
+
+From a user perspective, this feature doesn't introduce any breaking changes
+to existing deployments, because `status.hardware` field will stay for the
+time being and this feature is only an addition to the current APIs. For now,
+there will be two copies of inspection data available, in the `spec.hardware`
+of InspectionData and `status.hardware` of BareMetalHost.
```

Comment:
We should specify somewhere above how we plan to keep them in sync: i.e. copy one to the other if only one exists, new resource takes precedence if they are different.